clarity/src/vm/types/serialization.rs

```diff
@@ -452,275 +544,444 @@ impl Value {
+                            Value::list_with_type(
+                                &DESERIALIZATION_TYPE_CHECK_EPOCH,
+                                vec![],
+                                list_type.clone(),
+                            )
+                            .map_err(|_| "Illegal list type")?
+                        } else {
+                            Value::cons_list_unsanitized(vec![]).map_err(|_| "Illegal list type")?
+                        };
+
+                        Ok(finished_list)
                     }
                 }
+                TypePrefix::Tuple => {
+                    let mut len = [0; 4];
+                    r.read_exact(&mut len)?;
+                    let len = u32::from_be_bytes(len);
+                    let expected_len = u64::from(len);
+
+                    if len > MAX_VALUE_SIZE {
+                        return Err(SerializationError::DeserializationError(
+                            "Illegal tuple type".to_string(),
+                        ));
+                    }
 
-                let mut data = vec![0; u32::from(buffer_len) as usize];
-
-                r.read_exact(&mut data[..])?;
+                    let tuple_type = match expected_type.as_ref() {
+                        None => None,
+                        Some(TypeSignature::TupleType(tuple_type)) => {
+                            if sanitize {
+                                if u64::from(len) < tuple_type.len() {
+                                    return Err(SerializationError::DeserializeExpected(
+                                        expected_type.unwrap().clone(),
+                                    ));
+                                }
+                            } else {
+                                if len as u64 != tuple_type.len() {
+                                    return Err(SerializationError::DeserializeExpected(
+                                        expected_type.unwrap().clone(),
+                                    ));
+                                }
+                            }
+                            Some(tuple_type)
+                        }
+                        Some(x) => return Err(SerializationError::DeserializeExpected(x.clone())),
+                    };
 
-                Value::string_ascii_from_bytes(data).map_err(|_| "Bad string".into())
-            }
-            TypePrefix::StringUTF8 => {
-                let mut total_len = [0; 4];
-                r.read_exact(&mut total_len)?;
-                let total_len = BufferLength::try_from(u32::from_be_bytes(total_len))?;
+                    if len > 0 {
+                        let items = Vec::with_capacity(expected_len as usize);
+                        let first_key = ClarityName::deserialize_read(r)?;
+                        let next_sanitize = sanitize
+                            && tuple_type
+                                .map(|tt| tt.field_type(&first_key).is_none())
```

Comment:
It would be helpful to add a comment to explain this logic: 
`.map(|tt| tt.field_type(&first_key).is_none())`